src/core/or/sendme.c

```diff
@@ -555,16 +554,10 @@ sendme_note_cell_digest(circuit_t *circ)
     return;
   }
 
-  /* Only note the digest if we actually have the digest of the previous cell
-   * recorded. It should never happen in theory as we always record the last
-   * digest for the v1 SENDME. */
-  if (TO_OR_CIRCUIT(circ)->crypto.sendme_digest) {
-    digest = tor_malloc_zero(TRUNNEL_SENDME_V1_DIGEST_LEN);
-    crypto_digest_get_digest(TO_OR_CIRCUIT(circ)->crypto.sendme_digest,
-                             (char *) digest, TRUNNEL_SENDME_V1_DIGEST_LEN);
-    if (circ->sendme_last_digests == NULL) {
-      circ->sendme_last_digests = smartlist_new();
-    }
-    smartlist_add(circ->sendme_last_digests, digest);
+  /* Add the digest to the last seen list in the circuit. */
+  digest = TO_OR_CIRCUIT(circ)->crypto.sendme_digest;
```

Comment:
Let's have a function here instead that calculates the sendme digest for us from the relay_crypto.  That way:
  * we can get the digest through some different algorithm later on when other kinds of crypto are in use
  * we don't have to compute the digest with every cell
  * we don't have to make sendme.c know about the internals of relay_crypto.

Reply:
Not sure I understand. Probably here you mean a function *in* `relay_crypto.c` that returns the SENDME digest so we can use it to match the one in the circuit (`circ->sendme_last_digests`)?

Reply:
right -- I mean that we shouldn't reach inside the crypto object here. Instead we should have an accessor function.